Empire/Empire.cpp

```diff
@@ -1399,6 +1411,30 @@ void Empire::AddPartType(const std::string& name) {
         return;
     m_available_part_types.insert(name);
     AddSitRepEntry(CreateShipPartUnlockedSitRep(name));
+
+    // Decide whether to show the tutorial sitrep
+    // about building outpost or colony ships.
+    if (name == "CO_OUTPOST_POD") {
```

Comment:
The presence of content-specific strings in C++ code like this suggests the logic is misplaced. Whenever possible, put stuff like this into FOCS scripts, not hard-coded in C++.

Reply:
I totally see your point.  However, i did not find FOCS support for triggering Effects based on the *availability* of ship parts, and even less so edge-triggered by a ship-part becoming available.  Are you suggesting that i should implement scripting support for that, and if so, should it be "trigger effect when part *becomes* available" or "trigger effect when part *is* available" - in the latter case, it would have to be combined with "maxshow = 1", which may seem more rigorous and modular but less efficient, and right now, i don't see for which purpose one might want to *always* trigger an effect when a part is available.

Oh, maybe it would be more logical if a *technology* becoming available would edge-trigger effects, rather than a *part* becoming available?

And then, we need support for conditionals based on availability (of hulls or technologies) before it can be used for the intended purpose because this specific effect is intended to be triggered when (CO_OUTPOST_POD || CO_COLONY_POD) && (hull with Fuel() > 0 and NumSlots(SL_INTERNAL) > 0) are first available.

Reply:
You can already test the turn a tech was researched with the `TurnTechResearched` complex valueref.

I don't think it's necessary to specifically trigger against parts or hulls or similar. Tests against techs being researched should be sufficient, even if not exactly able to replicate the condition that any arbitrary other bit of content is being unlocked on a particular turn.

Reply:
@ischwarze, I want to chime in this discussion and point out the Python scripted turn events, if you haven't noticed them already. These are supposed to help scripting scenarios/campains, and also such things as tutorials. They should be more versatile/flexible for these kind of things than FOCS, and referencing content-specific string is completely fine there because the turn events are content too.

The exposed functionality is far from complete and can/should be expanded as need arises, e.g. when someone starts to work on things like in-game tutorials for real :wink:

Reply:
If these tutorial just-displayed-once effects can be implemented with Python turn event scripts, I'd much prefer that to the kludgy extra parameter...

Reply:
> If these tutorial just-displayed-once effects can be implemented with Python turn event scripts, I'd much prefer that to the kludgy extra parameter...

I'm working on that right now, and it looks promising.  I almost have the first tutorial sitrep working from turn_events.py (number 05: Introduction to Ship Design).  Once it really works, i'll push the required changes to the PR, and if you agree with the general direction, i can address the other ones.  Thanks to @Vezzra for the suggestion, it appears to be simpler, cleaner, and more powerful, all at the same time, compared to doing it in either C++ or pure FOCS.

Reply:
First step pushed in commit 2c813d9d (tutorial 05), and if that's what you meant, two or three more steps to go (03 and 06, and maybe also 07 to get rid of maxshow).  The others (01, 02, 04) are already pure FOCS.

Reply:
Last content-specific string removed from C++ code in commit f4d47c23, so i think this conversation can be resolved.